File: 0110_Balanced_Binary_Tree/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// T = O(n), S = O(n)
class Solution {
  public:
    bool isBalanced(TreeNode *root) { return dfs(root) != -1; }

    int dfs(TreeNode *root) {
        if (root == nullptr) {
            return 0;
        }

        int left = dfs(root->left);
        if (left == -1) {
            return -1;
        }

        int right = dfs(root->right);
        if (right == -1) {
            return -1;
        }

        if (abs(left - right) > 1) {
            return -1;
        }
        return max(left, right) + 1;
    }
};

File: 0110_Balanced_Binary_Tree/golang/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func max(a, b int) int {
	if a >= b {
		return a
	}
	return b
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func dfs(root *TreeNode) int {
	if root == nil {
		return 0
	}
	left := dfs(root.Left)
	if left == -1 {
		return -1
	}
	right := dfs(root.Right)
	if right == -1 {
		return -1
	}
	if abs(left-right) > 1 {
		return -1
	}
	return max(left, right) + 1
}

func isBalanced(root *TreeNode) bool {
	if root == nil {
		return true
	}
	return dfs(root) != -1
}